From 5dc9b1cf41357cb869657a2bd9c16955bfba65b6 Mon Sep 17 00:00:00 2001
From: Myron Stowe <myron.stowe@redhat.com>
Date: Fri, 5 Sep 2014 01:45:57 -0400
Subject: [iommu] amd: Fix PASID format in INVALIDATE_IOTLB_PAGES command

Message-id: <20140905014557.31297.70107.stgit@gir.stowe>
Patchwork-id: 91293
O-Subject: [RHEL7.1 PATCH 01/65] iommu/amd: Fix PASID format in INVALIDATE_IOTLB_PAGES command
Bugzilla: 1129880 1087643
RH-Acked-by: Stefan Assmann <sassmann@redhat.com>
RH-Acked-by: Alex Williamson <alex.williamson@redhat.com>
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>

Bugzilla: https://bugzilla.redhat.com/show_bug.cgi?id=1087643
Bugzilla: https://bugzilla.redhat.com/show_bug.cgi?id=1129880
Upstream Status: e8d2d82d4a73f37b3270e4fd19ba83e48b589656

commit e8d2d82d4a73f37b3270e4fd19ba83e48b589656
Author: Jay Cornwall <jay.cornwall@amd.com>
Date:   Wed Feb 26 15:49:31 2014 -0600

    iommu/amd: Fix PASID format in INVALIDATE_IOTLB_PAGES command

    This patch corrects the PASID format in the INVALIDATE_IOTLB_PAGES
    command, which was caused by incorrect information in
    the AMD IOMMU Architectural Specification v2.01 document.

        Incorrect format:
             cmd->data[0][16:23] = PASID[7:0]
             cmd->data[1][16:27] = PASID[19:8]

         Correct format:
             cmd->data[0][16:23] = PASID[15:8]
             cmd->data[1][16:23] = PASID[7:0]

    However, this does not affect the IOMMUv2 hardware implementation,
    and has been corrected since version 2.02 of the specification
    (available through AMD NDA).

    Signed-off-by: Jay Cornwall <jay.cornwall@amd.com>
    Reviewed-by: Suravee Suthikulpanit <suravee.suthikulpanit@amd.com>
    Signed-off-by: Joerg Roedel <joro@8bytes.org>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/iommu/amd_iommu.c b/drivers/iommu/amd_iommu.c
index 71776ff..724bf07 100644
--- a/drivers/iommu/amd_iommu.c
+++ b/drivers/iommu/amd_iommu.c
@@ -982,10 +982,10 @@ static void build_inv_iotlb_pasid(struct iommu_cmd *cmd, u16 devid, int pasid,
  address &= ~(0xfffULL);
 
  cmd->data[0]  = devid;
- cmd->data[0] |= (pasid & 0xff) << 16;
+ cmd->data[0] |= ((pasid >> 8) & 0xff) << 16;
  cmd->data[0] |= (qdep  & 0xff) << 24;
  cmd->data[1]  = devid;
- cmd->data[1] |= ((pasid >> 8) & 0xfff) << 16;
+ cmd->data[1] |= (pasid & 0xff) << 16;
  cmd->data[2]  = lower_32_bits(address);
  cmd->data[2] |= CMD_INV_IOMMU_PAGES_GN_MASK;
  cmd->data[3]  = upper_32_bits(address);
-- 
1.7.1